rust/operator-binary/src/spark_controller.rs

```diff
@@ -0,0 +1,968 @@
+
+    let master_role_service = build_master_role_service(
+        &sc,
+        default_master_role_ports
+            .iter()
+            .map(|(name, value)| ServicePort {
+                name: Some(name.clone()),
+                port: *value,
+                protocol: Some("TCP".to_string()),
+                ..ServicePort::default()
+            })
+            .collect(),
+    )?;
+
+    client
+        .apply_patch(
+            FIELD_MANAGER_SCOPE,
+            &master_role_service,
+            &master_role_service,
+        )
+        .await
+        .map_err(|e| ApplyRoleService {
+            source: e,
+            sc: sc_ref.clone(),
+        })?;
+
+    for (role_name, group_config) in validated_config.iter() {
+        for (rolegroup_name, rolegroup_config) in group_config.iter() {
+            let rolegroup = sc.server_rolegroup_ref(role_name, rolegroup_name);
+            let rg_service = build_rolegroup_service(&sc, &rolegroup, rolegroup_config)?;
+            let rg_configmap = build_rolegroup_config_map(&sc, &rolegroup, rolegroup_config)?;
+            let rg_statefulset = build_rolegroup_statefulset(
+                &sc,
+                &default_master_role_ports,
+                &rolegroup,
+                rolegroup_config,
+            )?;
+            client
+                .apply_patch(FIELD_MANAGER_SCOPE, &rg_service, &rg_service)
+                .await
+                .map_err(|e| ApplyRoleGroupService {
+                    source: e,
+                    rolegroup: rolegroup.clone(),
+                })?;
+            client
+                .apply_patch(FIELD_MANAGER_SCOPE, &rg_configmap, &rg_configmap)
+                .await
+                .map_err(|e| ApplyRoleGroupConfig {
+                    source: e,
+                    rolegroup: rolegroup.clone(),
+                })?;
+            client
+                .apply_patch(FIELD_MANAGER_SCOPE, &rg_statefulset, &rg_statefulset)
+                .await
+                .map_err(|e| ApplyRoleGroupStatefulSet {
+                    source: e,
+                    rolegroup: rolegroup.clone(),
+                })?;
+        }
+    }
+
+    Ok(ReconcilerAction {
+        requeue_after: None,
+    })
+}
+
+/// Build the [`NodePort`] service for clients.
+fn build_master_role_service(
+    sc: &SparkCluster,
+    service_ports: Vec<ServicePort>,
+) -> Result<Service, Error> {
+    let role_name = SparkRole::Master.to_string();
+    let role_svc_name = sc
+        .server_role_service_name()
+        .ok_or(GlobalServiceNameNotFound {
+            obj_ref: ObjectRef::from_obj(sc),
+        })?;
+    Ok(Service {
+        metadata: ObjectMetaBuilder::new()
+            .name_and_namespace(sc)
+            .name(&role_svc_name)
+            .ownerreference_from_resource(sc, None, Some(true))
+            .map_err(|e| ObjectMissingMetadataForOwnerRef {
+                source: e,
+                obj_ref: ObjectRef::from_obj(sc),
```

Comment:
`ObjectRef` can be removed from the errors. Will be logged implicitly.

Reply:
Unfortunately not because I'm not using snafu magic in this module.

Reply:
Ah true, it doesnt work if its properly separated :)